lib/agent.js

```diff
@@ -0,0 +1,80 @@
+'use strict'
+
+const { InvalidArgumentError, InvalidReturnValueError, NotSupportedError } = require('./core/errors')
+const { Pool } = require('../')
+const { weakCache, nodeMajorVersionIsGreaterThanOrEqualTo } = require('./core/util')
+
+const kAgentFactory = Symbol('agent factory')
+
+class Agent {
+  constructor (opts) {
+    /* istanbul ignore next */
+    if (!nodeMajorVersionIsGreaterThanOrEqualTo(14)) {
+      throw new NotSupportedError('Node.js version must be 14 or above')
+    }
+    this[kAgentFactory] = weakCache(origin => new Pool(origin, opts))
+  }
+
+  get (origin) {
+    if (typeof origin !== 'string' || origin === '') {
+      throw new InvalidArgumentError('Origin must be a non-empty string.')
+    }
+
+    return this[kAgentFactory](origin)
```

Comment:
Should we support unix socket also?

Reply:
nice to have

Reply:
The `socketPath` option should be used in `origin` to identify a pool of connections. Also, it should be passed to the `kAgentFactory`. It can be done in a new PR if it's not urgent.

Reply:
I'm +1 for follow up PR as this one is already quite large